solutions/joi-19-cake-circle.mdx

```diff
@@ -0,0 +1,180 @@
+---
+id: joi-19-cake-circle
+source: JOI 2019
+title: Cake Circle
+author: Dustin Miao
+---
+
+# Explanation
+
+First, given some set of cakes $S$, it is always optimal to arrange them in sorted order. The total color is $2(\max(S) - \min(S))$. 
+
+Sort the cakes by color. The task now becomes choosing some range $[l, r]$ that maximizes the sum of the $m$ largest beauty values, minus $2 \cdot (C_r - C_l)$.
+
+Let $\texttt{opt}(l)$ be the optimal right endpoint for some left endpoint. We can observe that $\texttt{opt}$ is a monotonically increasing function. This motivates a divide and conquer solution. 
+
+Suppose we know that $\texttt{opt}(i) \in [p, q]$ for $i \in [l, r]$. If $p = q$, then we are done. Otherwise, let $m = \lfloor \frac{l+r}{2}\rfloor$, and compute $\texttt{opt}(m)$. We now know that $\texttt{opt}(i)$ lies in the range $[p, \texttt{opt(m)}]$ for $i \in [l, m)$, and in the range $[\texttt{opt}(m), q]$ for $i \in (m, r]$. We can then recurse. By the [master theorem](https://en.wikipedia.org/wiki/Master_theorem_(analysis_of_algorithms)), this will run in $\tilde{\mathcal{O}}(n)$ time. 
+
+The only step left is to compute the sum of the $m$ largest values in a range. This can be done with a variety of data structures in logarithmic time.
+
+**Time Complexity**: $\mathcal{O}(n\log^2n)$
+
+# Implementation 
+
+<LanguageSection>
+
+<CPPSection>
+  
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+namespace std {
+ 
+template<class Fun>
+class y_combinator_result {
+	Fun fun_;
+public:
+	template<class T>
+	explicit y_combinator_result(T &&fun): fun_(std::forward<T>(fun)) {}
+ 
+	template<class ...Args>
+	decltype(auto) operator()(Args &&...args) {
+		return fun_(std::ref(*this), std::forward<Args>(args)...);
+	}
+};
+ 
+template<class Fun>
+decltype(auto) y_combinator(Fun &&fun) {
+	return y_combinator_result<std::decay_t<Fun>>(std::forward<Fun>(fun));
+}
+ 
+} // namespace std
```

Comment:
```suggestion
}  // namespace std
```

Reply:
Its formatted like this in the original source (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0200r0.html)

Reply:
it's just a single space